torch/csrc/utils/python_arg_parser.h

```diff
@@ -76,6 +76,51 @@
 #include <string>
 #include <vector>
 
+namespace torch {
+inline bool is_symint_node(py::handle obj) {
+  auto static tp_symn = py::type::of<c10::SymIntNodeImpl>();
+  // TODO: switch this to `isinstance`
+  if (obj.get_type().equal(tp_symn)) {
+    TORCH_CHECK(
+        !jit::tracer::isTracing(), "JIT tracing of SymInts isn't supported!");
+    return true;
+  }
+  return false;
+}
+} // namespace torch
+
+namespace pybind11 {
+namespace detail {
+template <>
+struct type_caster<c10::SymInt> {
+ public:
+  PYBIND11_TYPE_CASTER(c10::SymInt, const_name("SymInt"));
+  bool load(py::handle src, bool) {
+    if (torch::is_symint_node(src)) {
+      auto si = src.cast<c10::SymIntNodeImpl*>()->toSymInt();
+      value = si;
+      return true;
+    }
+
+    auto raw_obj = src.ptr();
+    if (THPUtils_checkIndex(raw_obj)) {
+      value = c10::SymInt{THPUtils_unpackIndex(raw_obj)};
```

Comment:
In a follow up PR, you should refactor the implicit constructor for SymInt to only accept integers that would be interpreted as integers (and not symbolic nodes), adding an alternate named constructor (use the enum trick to distinguish overloads) for setting the raw integer value in question. This will fix the latent bug here, which is that you will happily convert a negative integer that is too large into a pointer causing a potential crash later.